### PR TITLE
Converted potentially large variables to type long

### LINKED ISSUE #3228: Overflow in MultiClassNaiveBayes
We're storing the count of rows in an `int`. This causes an overflow for large datasets. In my case, Criteo 1TB w/ 4.4B rows. Recommend changing to a `long`.

https://github.com/dotnet/machinelearning/blob/d2bf3e72fa935b12644b0a695ad0d044a8077a83/src/Microsoft.ML.StandardTrainers/Standard/MulticlassClassification/MulticlassNaiveBayesTrainer.cs#L242

Error:
```
Unexpected exception: Arithmetic operation resulted in an overflow., 'System.OverflowException'
   at System.Linq.Enumerable.Sum(IEnumerable`1 source)
```

### PR DESCRIPTION
**Pull request by mstfbl**
Fixes #3228

As explained in Issue #3228, very large datasets more than 2.14 billion rows of data can cause overflow when, say, the sum of these labels are obtained, **and** if these are stored as ints. This PR converts arrays and matrices for storing labels and features in their respective histograms from type `int` to type `long`. In addition, this PR updates the version of NaiveBayesMulticlassTrainer's Loader to preserve backwards compatibility.

### CODE REVIEW

File: src/Microsoft.ML.StandardTrainers/Standard/MulticlassClassification/MulticlassNaiveBayesTrainer.cs [hunk 1]
```diff
@@ -294,25 +294,25 @@ private NaiveBayesMulticlassModelParameters(IHostEnvironment env, ModelLoadConte
             : base(env, LoaderSignature, ctx)
         {
             // *** Binary format ***
-            // int: _labelCount
-            // int[_labelCount]: _labelHistogram
+            // int: _labelCount (read during reading of _labelHistogram in ReadLongArray())
```
Comment (mstfbl) [inline src/Microsoft.ML.StandardTrainers/Standard/MulticlassClassification/MulticlassNaiveBayesTrainer.cs:297]: I added this comment as we are not explicitly reading _labelCount here, but in `ReadLongArray()` as shown below:

https://github.com/dotnet/machinelearning/blob/214926fcf5753ad62acf32c0759bdaf8fcd13b73/src/Microsoft.ML.Core/Utilities/Stream.cs#L681-L688

File: src/Microsoft.ML.StandardTrainers/Standard/MulticlassClassification/MulticlassNaiveBayesTrainer.cs [hunk 2]
```diff
@@ -339,22 +339,23 @@ private protected override void SaveCore(ModelSaveContext ctx)
...
             // int: _labelCount
-            // int[_labelCount]: _labelHistogram
+            // long[_labelCount]: _labelHistogram
             // int: _featureCount
-            // int[_labelCount][_featureCount]: _featureHistogram
+            // long[_labelCount][_featureCount]: _featureHistogram
             // int[_labelCount]: _absentFeaturesLogProb
-            ctx.Writer.WriteIntArray(_labelHistogram.AsSpan(0, _labelCount));
+            ctx.Writer.Write(_labelCount);
+            ctx.Writer.WriteLongStream(_labelHistogram);
```
Comment (mstfbl) [inline src/Microsoft.ML.StandardTrainers/Standard/MulticlassClassification/MulticlassNaiveBayesTrainer.cs:347]: `.AsSpan()` is not implemented for converting from 'System.Span<long>' to 'System.Collections.Generic.IEnumerable<long>. The only other way around this is using `Array.Copy()`, but that introduces needless array copying. I also don't see a case where not all of `_labelHistogram` is not serialized. As such, `_labelHistogram` is always serialized whole here.

File: src/Microsoft.ML.StandardTrainers/Standard/MulticlassClassification/MulticlassNaiveBayesTrainer.cs [hunk 3]
```diff
@@ -339,22 +339,23 @@ private protected override void SaveCore(ModelSaveContext ctx)
...
             ctx.Writer.Write(_featureCount);
             for (int i = 0; i < _labelCount; i += 1)
             {
                 if (_labelHistogram[i] > 0)
-                    ctx.Writer.WriteIntsNoCount(_featureHistogram[i].AsSpan(0, _featureCount));
+                    ctx.Writer.WriteLongStream(_featureHistogram[i]);
```
Comment (mstfbl) [inline src/Microsoft.ML.StandardTrainers/Standard/MulticlassClassification/MulticlassNaiveBayesTrainer.cs:352]: Sames as above.

`.AsSpan()` is not implemented for converting from 'System.Span<long>' to 'System.Collections.Generic.IEnumerable<long>. The only other way around this is using `Array.Copy()`, but that introduces needless array copying. I also don't see a case where not all of `_labelHistogram` is not serialized. As such, `_labelHistogram` is always serialized whole here.

File: src/Microsoft.ML.StandardTrainers/Standard/MulticlassClassification/MulticlassNaiveBayesTrainer.cs [hunk 4]
```diff
@@ -294,25 +309,33 @@ private NaiveBayesMulticlassModelParameters(IHostEnvironment env, ModelLoadConte
...
-            // int[_labelCount][_featureCount]: _featureHistogram
+            // long[_labelCount][_featureCount]: _featureHistogram
             // int[_labelCount]: _absentFeaturesLogProb
-            _labelHistogram = ctx.Reader.ReadIntArray() ?? new int[0];
+            if (ctx.Header.ModelVerWritten >= 0x00010002)
+                _labelHistogram = ctx.Reader.ReadLongArray() ?? new long[0];
+            else
+            {
+                _labelHistogram = Array.ConvertAll(ctx.Reader.ReadIntArray() ?? new int[0], x => (long)x);
+            }
```
Comment (harishsk) [inline src/Microsoft.ML.StandardTrainers/Standard/MulticlassClassification/MulticlassNaiveBayesTrainer.cs:322]: If ReadIntArray returns null, it likely means the file is bad. Should you be throwing an error in this case?

File: src/Microsoft.ML.StandardTrainers/Standard/MulticlassClassification/MulticlassNaiveBayesTrainer.cs [hunk 5]
```diff
@@ -294,25 +309,33 @@ private NaiveBayesMulticlassModelParameters(IHostEnvironment env, ModelLoadConte
...
             for (int iLabel = 0; iLabel < _labelCount; iLabel += 1)
             {
                 if (_labelHistogram[iLabel] > 0)
                 {
-                    _featureHistogram[iLabel] = ctx.Reader.ReadIntArray(_featureCount);
+                    if (ctx.Header.ModelVerWritten >= 0x00010002)
+                        _featureHistogram[iLabel] = ctx.Reader.ReadLongArray(_featureCount);
+                    else
+                        _featureHistogram[iLabel] = Array.ConvertAll(ctx.Reader.ReadIntArray(_featureCount) ?? new int[0], x => (long)x);
                     for (int iFeature = 0; iFeature < _featureCount; iFeature += 1)
```
Comment (harishsk) [inline src/Microsoft.ML.StandardTrainers/Standard/MulticlassClassification/MulticlassNaiveBayesTrainer.cs:339]: Same comment as above

File: src/Microsoft.ML.StandardTrainers/Standard/MulticlassClassification/MulticlassNaiveBayesTrainer.cs (hunk 4 above)
Comment (mstfbl) [reply src/Microsoft.ML.StandardTrainers/Standard/MulticlassClassification/MulticlassNaiveBayesTrainer.cs:322]: Hey Harish, the array being read from `ctx.Reader.ReadIntArray(int size)` can return null if the size of the array being loaded is 0. Here's the source code:

https://github.com/dotnet/machinelearning/blob/7628d6cd06a0cf8294ddeebbb9613aaec72ab4e7/src/Microsoft.ML.Core/Utilities/Stream.cs#L605-L641

**Pull request merged**